Improve type-checking

InnocentZero opened the pull request:
Adds type-checking to the following cases:

- block outputs (reflect the type of the last statement, none otherwise)
- parameters in def (any if not specified, else the specified type)
- if else (reflect the type of the blocks if the blocks are equal, if they are unequal then forbidden, and the if branch type otherwise)

InnocentZero commented:
Updates:

- `Params` now has a `None` type, since they are not expected to have types anyways.
- `Block` and `BLOCK_TYPE` are now redundant, since blocks return the type of the last statement

kubouch commented:
About the if-else, if they have different types, it should be allowed and the type should be `oneof<a, b>`. For example,
```
if $cond { 4 } else { 'a' }
```
should have type `oneof<int, string>`. We can start just by casting it to `any`, but it definitely shouldn't throw an error.

Params having None type sounds good. Block having the last type also seems good.

InnocentZero commented:
Yeah, I realised that much later while going through docs and messing around in the shell, but that'd require some deeper changes. Anyways, I'll revert those changes in the next commit and hopefully add some more checks.

InnocentZero commented:
Updates:
- Removed the if-else checks since they were basically wrong. Now it only checks for the condition returning a boolean
- Added a `OneOf(...)` type that stores an index. We use that index in `Typechecker::oneof_types` to get the corresponding BTreeSet of types.
- if-else return type now supports `OneOf` if there are more that one types being returned. Else it will return the singleton type
- for loop and while loop typechecking has been added
- Added an alias test because I wanted to see the behaviour of the call node
- Typechecking for match node is now supported. Tests, however, suggest that it's not being parsed properly atm.

InnocentZero commented:
@kubouch should I attempt to fix match parsing or would you merge it rn? The PR looks decently sized and implements a couple of major changes, so I'd want to get a review if possible.

InnocentZero commented:
@kubouch would you like to review and merge it now or should I go on to fix match parsing/implement typechecking for other nodes?

kubouch commented:
Let's not add too much stuff to it, the match parsing can be left for another PR.

The summary seems good, I'll try to take a look whenever possible. Any specific reason to use BTreeSet over a plain Vec?

InnocentZero commented:
BTreeSet makes it easier to keep unique elements. I did not want a `oneof<int, int>` kind of situation.

kubouch commented:
I see, yeah, a set makes sense in this context. Any reason for the BTreeSet over a regular HashSet?

InnocentZero commented:
My bad, I was working on something else that required BTreeSet and forgot that HashSet exists when I switched back to working on this.

InnocentZero commented:
I'm not sure what to make of the CI tests :/

kubouch commented on tests/alias.nu:
Since alias is not in the parser yet, let's not add this test for now.

kubouch commented on src/snapshots/new_nu_parser__test__node_output@for_break_continue.nu.snap:
```diff
@@ -47,34 +48,35 @@ input_file: tests/for_break_continue.nu
-16: unknown
-17: unknown
-18: unknown
+14: oneof<(), unknown>
```
The type of missing `if` branch is `nothing`. So, `if $spam { 1 }` should have a return type of `oneof<int, nothing>`.

InnocentZero replied:
That `()` is from the else branch. The if branch does not have a type because continue and break are not checked/parsed properly.

kubouch replied:
Yeah, I meant the else branch. It should be `nothing`, not `()` coming from the missing else branch. Try `if false { 1 } | describe` in Nushell, it'll print `nothing`, so we should match the behavior.

kubouch commented on src/snapshots/new_nu_parser__test__node_output@binary_ops_mismatch.nu.snap:
```diff
@@ -32,7 +33,7 @@ input_file: tests/binary_ops_mismatch.nu
 10: string
 11: unknown
-12: ()
+12: unknown
```
Maybe `any` type would be better in case of error. The intent is that after compiling, there wouldn't be any `unknown` types anymore. `unknown` type means "untouched by a typechecker". That's the reason the `forbidden` type exists: To mark nodes visited by the typechecker that shouldn't be directly evaluated, like operators. Not sure if we need to stick with this that strictly, but that's how I've been thinking about it.

kubouch replied:
Another option would be to introduce an `error` type.

InnocentZero replied:
I think it's better to not have multiple `Any` interpretations. Let's introduce an error type. 

My intuition for leaving it untouched was that the compiler is going to throw an error anyways but I do see how that info might be useful.

kubouch replied:
Yeah, the `error` type is better, and it would match the `garbage` AST node nicely.

kubouch commented on src/snapshots/new_nu_parser__test__node_output@invalid_types.nu.snap:
```diff
@@ -44,22 +45,22 @@ input_file: tests/invalid_types.nu
 5: unknown
 6: string
-7: unknown
+7: ()
```
I'm wondering if Params should be `forbidden`. They should never be directly evaluated.

InnocentZero replied:
Yeah that makes sense. As long as we're not doing type *inferencing* we should be good

kubouch commented on src/snapshots/new_nu_parser__test__node_output@invalid_types.nu.snap:
```diff
@@ -44,22 +45,22 @@ input_file: tests/invalid_types.nu
-10: unknown
-11: unknown
-12: ()
+9: list<unknown>
```
I think Param should also be `forbidden` since it won't be directly evaluated?

InnocentZero replied:
I was thinking we can do some better typechecking by checking each param and if it is type compatible to the argument of the function once we have the facilities for that.

kubouch commented on src/typechecker.rs:
```diff
@@ -298,6 +349,79 @@ impl<'a> Typechecker<'a> {
+            } => {
+                self.typecheck_node(range);
+
+                // We don't need to typecheck variable after this
```
Why not? (The comment is not very descriptive.)

InnocentZero replied:
My bad. Basically the variable's is resolved when we typecheck the list. I think the comment is positioned wrong.

InnocentZero replied:
Not to mention, the variable's type is being set by us after checking the type of the range. We're setting the type in the for node itself, not before.

kubouch commented on src/snapshots/new_nu_parser__test__node_output@match.nu.snap:
```diff
@@ -0,0 +1,49 @@
+Error (NodeId 26): incomplete expression
+Error (NodeId 27): incomplete expression
+Error (NodeId 28): expected: right paren ')'
+Error (NodeId 30): incomplete expression
```
Seems like it's cropped?

InnocentZero replied:
I removed and ran cargo insta, seems like the output is the same.

kubouch commented on src/typechecker.rs:
```diff
@@ -298,6 +349,79 @@ impl<'a> Typechecker<'a> {
+                    self.error("Blocks in looping constructs cannot return values", block);
+                }
+
+                self.set_node_type_id(node_id, NOTHING_TYPE);
```
To match the current behavior of Nushell (i.e., `for i in 0..8 { print $i } | describe` throwing an error because `for` is a statement), we should probably make it into a `none` type.

However, in the future, we could perhaps allow `for` to be an expression, and you'd be able to break a value from it, e.g., `let x = for i in 1..8 { if $ > 3 { break ($i * $i) } }`. In that case, `nothing` type is appropriate for a loop without value breaks. This applies to `while` and `loop` as well.

InnocentZero replied:
For the second part, would it not be a problem if the for loop never evaluates with a list at runtime? Or do you plan to keep it `oneof<nothing, type>`?

InnocentZero replied:
For now I've gone with `none`

kubouch commented on src/typechecker.rs:
```diff
@@ -298,6 +349,79 @@ impl<'a> Typechecker<'a> {
+                        );
+                    }
+                    Ordering::Less => {
+                        self.set_node_type_id(node_id, NONE_TYPE);
```
Since `match` is an expression, I think this should be `nothing` because you can assign expressions to values. `none` belongs to statements.

kubouch commented:
OK, some minor comments, but otherwise looks good.

kubouch commented:
OK, let's try it, thanks!

I think the `match` parsing might be borked and you did some fixes on it in your other branch, so let's roll with it.
